File: src/test/java/integration/cardinstance/cost/PayCostServiceTest.java
package integration.cardinstance.cost;

import com.matag.cards.Cards;
import com.matag.cards.CardsConfiguration;
import com.matag.game.cardinstance.CardInstance;
import com.matag.game.cardinstance.CardInstanceFactory;
import com.matag.game.cardinstance.cost.PayCostService;
import com.matag.game.status.GameStatus;
import com.matag.game.turn.action.tap.TapPermanentService;
import integration.TestUtils;
import integration.TestUtilsConfiguration;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.util.List;
import java.util.Map;

import static org.mockito.Mockito.verify;

@ExtendWith(SpringExtension.class)
@Import({CardsConfiguration.class, TestUtilsConfiguration.class})
public class PayCostServiceTest {
  private int cardInstanceId = 60;

  @Autowired
  private TestUtils testUtils;

  @Autowired
  private Cards cards;

  @Autowired
  private CardInstanceFactory cardInstanceFactory;

  @Autowired
  private PayCostService payCostService;

  @Autowired
  private TapPermanentService tapPermanentService;

  @Test
  public void isCastingCostFulfilledFeralMaakaCorrectCosts() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = createCardInstance(gameStatus, "Feral Maaka");
    var mountain1 = createCardInstance(gameStatus, "Mountain");
    var mountain2 = createCardInstance(gameStatus, "Mountain");

    var manaPaid = Map.of(
      mountain1.getId(), List.of("RED"),
      mountain2.getId(), List.of("RED")
    );

    // When
    payCostService.pay(gameStatus, gameStatus.getActivePlayer(), cardInstance, null, manaPaid);

    // Then
    verify(tapPermanentService).tap(gameStatus, mountain1.getId());
    verify(tapPermanentService).tap(gameStatus, mountain2.getId());
  }

  @Test
  public void isCastingCostFulfilledCheckpointOfficerTapAbility() {
    // Given
    var gameStatus = testUtils.testGameStatus();
    var cardInstance = createCardInstance(gameStatus, "Checkpoint Officer");
    var plains1 = createCardInstance(gameStatus, "Plains");
    var plains2 = createCardInstance(gameStatus, "Plains");
    var manaPaid = Map.of(
      plains1.getId(), List.of("WHITE"),
      plains2.getId(), List.of("WHITE")
    );

    // When
    payCostService.pay(gameStatus, gameStatus.getActivePlayer(), cardInstance, "THAT_TARGETS_GET", manaPaid);

    // Then
    verify(tapPermanentService).tap(gameStatus, plains1.getId());
    verify(tapPermanentService).tap(gameStatus, plains2.getId());
    verify(tapPermanentService).tap(gameStatus, cardInstance.getId());
  }

  private CardInstance createCardInstance(GameStatus gameStatus, String cardName) {
    var card = cards.get(cardName);
    var cardInstance = cardInstanceFactory.create(gameStatus, ++cardInstanceId, card, "player-name", "player-name");
    gameStatus.getActivePlayer().getBattlefield().addCard(cardInstance);
    return cardInstance;
  }
}

File: src/test/kotlin/integration/cardinstance/cost/PayCostServiceTest.kt
package integration.cardinstance.cost

import com.matag.game.MatagGameApplication
import com.matag.game.cardinstance.cost.PayCostService
import com.matag.game.status.GameStatus
import integration.TestUtils
import integration.TestUtilsConfiguration
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles

@SpringBootTest(classes = [MatagGameApplication::class], webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ActiveProfiles("test")
@Import(TestUtilsConfiguration::class)
class PayCostServiceTest(
    val testUtils: TestUtils,
    val payCostService: PayCostService
) {
    @Test
    fun `pay for casting Feral Maaka`() {
        // Given
        val gameStatus = testUtils.testGameStatus()
        val cardInstance = addCardToPlayerBattlefield(gameStatus, "Feral Maaka")
        val mountain1 = addCardToPlayerBattlefield(gameStatus, "Mountain")
        val mountain2 = addCardToPlayerBattlefield(gameStatus, "Mountain")

        val manaPaid = mapOf(
            mountain1.id to listOf("RED"),
            mountain2.id to listOf("RED")
        )

        // When
        payCostService.pay(gameStatus, gameStatus.activePlayer, cardInstance, null, manaPaid)

        // Then
        assertThat(mountain1.modifiers.isTapped).isTrue
        assertThat(mountain2.modifiers.isTapped).isTrue
    }

    @Test
    fun `pay for casting tap target creation ability of Checkpoint Officer`() {
        // Given
        val gameStatus = testUtils.testGameStatus()
        val cardInstance = addCardToPlayerBattlefield(gameStatus, "Checkpoint Officer")
        val plains1 = addCardToPlayerBattlefield(gameStatus, "Plains")
        val plains2 = addCardToPlayerBattlefield(gameStatus, "Plains")
        val manaPaid = mapOf(
            plains1.id to listOf("WHITE"),
            plains2.id to listOf("WHITE")
        )

        // When
        payCostService.pay(gameStatus, gameStatus.activePlayer, cardInstance, "THAT_TARGETS_GET", manaPaid)

        // Then
        assertThat(plains1.modifiers.isTapped).isTrue
        assertThat(plains2.modifiers.isTapped).isTrue
        assertThat(cardInstance.modifiers.isTapped).isTrue
    }

    private fun addCardToPlayerBattlefield(gameStatus: GameStatus, cardName: String) =
        testUtils.createCardInstance(gameStatus, cardName).also {
            gameStatus.activePlayer.battlefield.addCard(it)
        }
}